File: Provincial_Excellent_Student_Contest/Binh_Dinh/Secondary_school/2023/Solution/three.py
from array import array
import math

def solve():
    a = array('u', sorted(s))
    count = array('i', [])
    pos = array('i', [])
    i = 0
    while i < len(a):
        j = i
        while j + 1 < len(a):
            if a[i] == a[j + 1]:
                j += 1
            else:
                break
        pos.append(i)
        count.append(j - i + 1)
        i = j + 1
    
    if len(pos) == 1: 
        subtask = 0
    else: 
        subtask = 1

    res = array('u', [])

    if subtask == 0:
        for _ in range(math.ceil(n / k)):
            res.append(a[0])
            
    if subtask == 1:
        if count[0] < k:
            res.append(a[k-1])
        else:
            if count[0] > k:
                for _ in range(count[0] - k + 1):
                    res.append(a[0])
                for i in range(pos[1], len(a)):
                    res.append(a[i])
            else:   # count[0] == k;
                res.append(a[0])
                i = 0
                while (i + 1 < len(count)):
                    i += 1
                    if count[i] % k == 0:
                        for _ in range(count[i] // k):
                            res.append(a[pos[i]])
                    else: 
                        for j in range(pos[i], len(a)):
                            res.append(a[j])
                        break
            
    # print(pos)                    
    # print(count)
    print(''.join(x for x in res))
        
        
                





if __name__ == '__main__':
    n, k = map(int, input().split())
    s = input()
    solve()

File: Provincial_Excellent_Student_Contest/Binh_Dinh/Secondary_school/2023/Solution/three.pas
program Solve;
uses math, sysutils;

var
  a, count, pos: array of Integer;
  res: string;
  i, j, subtask, n, k: Integer;
  s: string;

procedure QuickSort(var a: array of Char; left, right: Integer);
var
  i, j, pivot: Integer;
  temp: Char;
begin
  if left < right then
  begin
    pivot := left;
    i := left;
    j := right;
    while i < j do
    begin
      while (a[i] <= a[pivot]) and (i < right) do
        i := i + 1;
      while a[j] > a[pivot] do
        j := j - 1;
      if i < j then
      begin
        temp := a[i];
        a[i] := a[j];
        a[j] := temp;
      end;
    end;
    temp := a[pivot];
    a[pivot] := a[j];
    a[j] := temp;
    QuickSort(a, left, j - 1);
    QuickSort(a, j + 1, right);
  end;
end;

procedure Solve;
var
  i: Integer;
begin
  SetLength(a, Length(s));
  for i := 1 to Length(s) do
    a[i-1] := Ord(s[i]);
  QuickSort(a, 0, Length(a) - 1);

  i := 0;
  while i < Length(a) do
  begin
    j := i;
    while (j + 1 < Length(a)) and (a[i] = a[j+1]) do
      j := j + 1;
    SetLength(pos, Length(pos) + 1);
    pos[Length(pos)-1] := i;
    SetLength(count, Length(count) + 1);
    count[Length(count)-1] := j - i + 1;
    i := j + 1;
  end;

  if Length(pos) = 1 then 
    subtask := 0
  else 
    subtask := 1;

  res := '';

  if subtask = 0 then
  begin
    for i := 1 to Ceil(n / k) do
      res := res + Char(a[0]);
  end;

  if subtask = 1 then
  begin
    if count[0] < k then
      res := res + Char(a[k-1])
    else
    begin
      if count[0] > k then
      begin
        for i := 1 to count[0] - k do
          res := res + Char(a[0]);
        for i := pos[1] to Length(a) - 1 do
          res := res + Char(a[i]);
      end
      else  // count[0] = k;
      begin
        res := res + Char(a[0]);
        i := 0;
        while (i + 1 < Length(count)) do
        begin
          i := i + 1;
          if count[i] mod k = 0 then
            for j := 1 to count[i] div k do
              res := res + Char(a[pos[i]])
          else
          begin
            for j := pos[i] to Length(a) - 1 do
              res := res + Char(a[j]);
            break;
          end;
        end;
      end;
    end;

    WriteLn(res);
  end;
end;

begin
  ReadLn(n, k);
  ReadLn(s);
  Solve;
end."
